src/topology/schema.rs

```diff
@@ -30,82 +31,59 @@ use crate::{
 ///
 /// Finally, The merged definition (named `Definition 1 & 2`), and `Definition 4` are merged
 /// together to produce the new `Definition` returned by this method.
-pub(super) fn merged_definition(
+pub fn merged_definition(
     inputs: &[OutputId],
     config: &dyn ComponentContainer,
-    cache: &mut HashMap<Vec<OutputId>, Definition>,
+    cache: &mut HashMap<(bool, Vec<OutputId>), Definition>,
 ) -> Definition {
     // Try to get the definition from the cache.
-    if let Some(definition) = cache.get(inputs) {
+    if let Some(definition) = cache.get(&(config.schema_enabled(), inputs.to_vec())) {
         return definition.clone();
     }
 
-    // This should be "never", but that isn't supported yet (coming soon!)
-    let mut definition: Option<Definition> = None;
+    // TODO: special case when inputs is empty?
+    let mut definition = Definition::empty_kind(Kind::never(), []);
 
     for input in inputs {
         let key = &input.component;
 
-        // If the input is a source, it'll always have schema definition attached, even if it is an
-        // "empty" schema.
-        //
-        // We merge this schema into the top-level schema.
-        if let Some(outputs) = config.source_outputs(key) {
-            // After getting the source matching to the given input, we need to further narrow the
-            // actual output of the source feeding into this input, and then get the definition
-            // belonging to that output.
-            let maybe_source_definition = outputs.iter().find_map(|output| {
-                if output.port == input.port {
-                    // For sources, a `None` schema definition is equal to an "empty" definition.
-                    Some(
-                        output
-                            .log_schema_definition
-                            .clone()
-                            .unwrap_or_else(Definition::empty),
+        // Merge the schema if it's a source
+        if let Ok(maybe_output) = config.source_output_for_port(key, &input.port) {
+            let source_definition = maybe_output
+                .unwrap_or_else(|| {
+                    panic!(
+                        "source output mis-configured - output for port {:?} missing",
+                        &input.port
                     )
```

Comment:
Given that the original code mentions this can never happen (i.e. other parts of the code guard against this), I think I still prefer an `unreachable` here.